File: renderweave-inference/src/main/java/cn/hbads/renderweave/inference/eval/visual/quality/R5P2SourceLineReconciliation.java
package cn.hbads.renderweave.inference.eval.visual.quality;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.HexFormat;
import java.util.List;
import java.util.Objects;

/** Pure evaluation-memory source-line reconciliation for R5P2. */
public final class R5P2SourceLineReconciliation {
    public static final String VERSION = "FrozenSourceLineReconciliationPolicy/1.0";
    public static final String PROJECTION_IDENTITY = "renderweave-r5p-source-projection/1.0";
    public static final int AREA_OVERLAP_BPS = 5_000;
    public static final int VERTICAL_OVERLAP_BPS = 8_000;
    private static final List<String> POLICY_FRAMES = List.of(
            VERSION,
            "projection=" + PROJECTION_IDENTITY,
            "scope=same-source-cross-view-only/1.0",
            "geometry=intersection-over-smaller-area-5000-bps/1.0",
            "vertical=intersection-over-smaller-height-8000-bps/1.0",
            "center=smaller-center-in-larger-closed-open/1.0",
            "cluster=complete-link-source-order/1.0",
            "representative=pixel-density-confidence-smaller-area-view-line/1.0",
            "representative-payload=original-observation-only/1.0",
            "canonical-text-order=unicode-scalar-lexicographic/1.0",
            "sort=top-left-bottom-right-canonical-text-view-line/1.0");
    public static final String POLICY_IDENTITY = VERSION + ":" + sha256(
            String.join("\n", POLICY_FRAMES).getBytes(StandardCharsets.UTF_8));

    private static final Comparator<ProjectedLine> SOURCE_ORDER = Comparator
            .comparing((ProjectedLine value) -> value.sourceBox().top())
            .thenComparing(value -> value.sourceBox().left())
            .thenComparing(value -> value.sourceBox().bottom())
            .thenComparing(value -> value.sourceBox().right())
            .thenComparing(ProjectedLine::text, R5P2SourceLineReconciliation::compareUnicodeScalars)
            .thenComparingInt(ProjectedLine::viewOrdinal)
            .thenComparingInt(ProjectedLine::lineOrdinal);

    private R5P2SourceLineReconciliation() { }

    public static ProjectedLine project(
            String observationId,
            String sourceArtifactId,
            int viewOrdinal,
            int lineOrdinal,
            int viewWidth,
            int viewHeight,
            int sourceWidth,
            int sourceHeight,
            PixelBox sourceCrop,
            PixelBox viewLine,
            int confidenceBps,
            String text
    ) {
        Objects.requireNonNull(sourceCrop, "sourceCrop").requireWithin(sourceWidth, sourceHeight);
        Objects.requireNonNull(viewLine, "viewLine").requireWithin(viewWidth, viewHeight);
        var viewCanonical = new SourceBox(
                floorRatio(viewLine.left(), 10_000, viewWidth),
                floorRatio(viewLine.top(), 10_000, viewHeight),
                ceilRatio(viewLine.right(), 10_000, viewWidth),
                ceilRatio(viewLine.bottom(), 10_000, viewHeight));
        var cropWidth = sourceCrop.right() - sourceCrop.left();
        var cropHeight = sourceCrop.bottom() - sourceCrop.top();
        var sourceBox = new SourceBox(
                projectFloor(sourceCrop.left(), cropWidth, viewCanonical.left(), sourceWidth),
                projectFloor(sourceCrop.top(), cropHeight, viewCanonical.top(), sourceHeight),
                projectCeil(sourceCrop.left(), cropWidth, viewCanonical.right(), sourceWidth),
                projectCeil(sourceCrop.top(), cropHeight, viewCanonical.bottom(), sourceHeight));
        return new ProjectedLine(observationId, sourceArtifactId, sourceBox, confidenceBps,
                text, viewOrdinal, lineOrdinal,
                new PixelDensity(Math.multiplyExact((long) viewWidth, viewHeight),
                        Math.multiplyExact((long) cropWidth, cropHeight)));
    }

    public static Outcome reconcile(List<ProjectedLine> input) {
        input = List.copyOf(Objects.requireNonNull(input, "input"));
        if (input.isEmpty() || input.size() > 4_096 || input.stream().anyMatch(Objects::isNull)) {
            throw invalid("R5P2_RECONCILIATION_INPUT_INVALID");
        }
        var observationIds = new HashSet<String>();
        for (var line : input) {
            if (!observationIds.add(line.observationId())) {
                throw invalid("R5P2_RECONCILIATION_OBSERVATION_DUPLICATED");
            }
        }
        var ordered = new ArrayList<>(input);
        ordered.sort(SOURCE_ORDER);
        var clusters = new ArrayList<List<ProjectedLine>>();
        for (var candidate : ordered) {
            var assigned = false;
            for (var cluster : clusters) {
                if (cluster.stream().allMatch(existing -> sameSourceLineCandidate(candidate, existing))) {
                    cluster.add(candidate);
                    assigned = true;
                    break;
                }
            }
            if (!assigned) {
                var cluster = new ArrayList<ProjectedLine>();
                cluster.add(candidate);
                clusters.add(cluster);
            }
        }
        var representatives = new ArrayList<ProjectedLine>();
        for (var cluster : clusters) {
            var representative = cluster.getFirst();
            for (var candidate : cluster) {
                if (prefersRepresentative(candidate, representative)) representative = candidate;
            }
            representatives.add(representative);
        }
        representatives.sort(SOURCE_ORDER);
        return new Outcome(List.copyOf(representatives), input.size(), clusters.size(), POLICY_IDENTITY);
    }

    public static boolean sameSourceLineCandidate(ProjectedLine left, ProjectedLine right) {
        Objects.requireNonNull(left, "left");
        Objects.requireNonNull(right, "right");
        if (!left.sourceArtifactId().equals(right.sourceArtifactId())
                || left.viewOrdinal() == right.viewOrdinal()) {
            return false;
        }
        var leftBox = left.sourceBox();
        var rightBox = right.sourceBox();
        var intersectionWidth = Math.max(0,
                Math.min(leftBox.right(), rightBox.right()) - Math.max(leftBox.left(), rightBox.left()));
        var intersectionHeight = Math.max(0,
                Math.min(leftBox.bottom(), rightBox.bottom()) - Math.max(leftBox.top(), rightBox.top()));
        var intersection = Math.multiplyExact((long) intersectionWidth, intersectionHeight);
        var smallerArea = Math.min(leftBox.area(), rightBox.area());
        var smallerHeight = Math.min(leftBox.height(), rightBox.height());
        if (intersection == 0
                || !areaThresholdAllows(intersection, smallerArea)
                || !verticalThresholdAllows(intersectionHeight, smallerHeight)) {
            return false;
        }
        if (leftBox.area() < rightBox.area()) return centerInside(leftBox, rightBox);
        if (rightBox.area() < leftBox.area()) return centerInside(rightBox, leftBox);
        return centerInside(leftBox, rightBox) && centerInside(rightBox, leftBox);
    }

    public static boolean areaThresholdAllows(long intersectionArea, long smallerArea) {
        return atLeastBps(intersectionArea, smallerArea, AREA_OVERLAP_BPS);
    }

    public static boolean verticalThresholdAllows(long intersectionHeight, long smallerHeight) {
        return atLeastBps(intersectionHeight, smallerHeight, VERTICAL_OVERLAP_BPS);
    }

    public static boolean prefersRepresentative(ProjectedLine candidate, ProjectedLine existing) {
        Objects.requireNonNull(candidate, "candidate");
        Objects.requireNonNull(existing, "existing");
        var density = candidate.pixelDensity().compareTo(existing.pixelDensity());
        if (density != 0) return density > 0;
        var confidence = Integer.compare(candidate.confidenceBps(), existing.confidenceBps());
        if (confidence != 0) return confidence > 0;
        var area = Long.compare(candidate.sourceBox().area(), existing.sourceBox().area());
        if (area != 0) return area < 0;
        if (candidate.viewOrdinal() != existing.viewOrdinal()) {
            return candidate.viewOrdinal() < existing.viewOrdinal();
        }
        return candidate.lineOrdinal() < existing.lineOrdinal();
    }

    private static int compareUnicodeScalars(String left, String right) {
        var leftOffset = 0;
        var rightOffset = 0;
        while (leftOffset < left.length() && rightOffset < right.length()) {
            var leftCodePoint = left.codePointAt(leftOffset);
            var rightCodePoint = right.codePointAt(rightOffset);
            if (leftCodePoint != rightCodePoint) {
                return Integer.compare(leftCodePoint, rightCodePoint);
            }
            leftOffset += Character.charCount(leftCodePoint);
            rightOffset += Character.charCount(rightCodePoint);
        }
        if (leftOffset == left.length()) return rightOffset == right.length() ? 0 : -1;
        return 1;
    }

    private static boolean centerInside(SourceBox smaller, SourceBox larger) {
        var centerX2 = Math.addExact(smaller.left(), smaller.right());
        var centerY2 = Math.addExact(smaller.top(), smaller.bottom());
        return Math.multiplyExact(larger.left(), 2) <= centerX2
                && centerX2 < Math.multiplyExact(larger.right(), 2)
                && Math.multiplyExact(larger.top(), 2) <= centerY2
                && centerY2 < Math.multiplyExact(larger.bottom(), 2);
    }

    private static boolean atLeastBps(long numerator, long denominator, int threshold) {
        return denominator > 0 && Math.multiplyExact(numerator, 10_000L)
                >= Math.multiplyExact(denominator, (long) threshold);
    }

    private static int floorRatio(int value, int multiplier, int divisor) {
        if (divisor < 1) throw invalid("R5P2_PROJECTION_DIMENSIONS_INVALID");
        return Math.toIntExact(Math.floorDiv(Math.multiplyExact((long) value, multiplier), divisor));
    }

    private static int ceilRatio(int value, int multiplier, int divisor) {
        if (divisor < 1) throw invalid("R5P2_PROJECTION_DIMENSIONS_INVALID");
        return Math.toIntExact(Math.ceilDiv(Math.multiplyExact((long) value, multiplier), divisor));
    }

    private static int projectFloor(int cropStart, int cropSize, int coordinate, int sourceSize) {
        if (sourceSize < 1) throw invalid("R5P2_PROJECTION_DIMENSIONS_INVALID");
        return Math.toIntExact(Math.floorDiv(Math.addExact(
                Math.multiplyExact((long) cropStart, 10_000L),
                Math.multiplyExact((long) coordinate, cropSize)), sourceSize));
    }

    private static int projectCeil(int cropStart, int cropSize, int coordinate, int sourceSize) {
        if (sourceSize < 1) throw invalid("R5P2_PROJECTION_DIMENSIONS_INVALID");
        return Math.toIntExact(Math.ceilDiv(Math.addExact(
                Math.multiplyExact((long) cropStart, 10_000L),
                Math.multiplyExact((long) coordinate, cropSize)), sourceSize));
    }

    private static String canonicalText(String value) {
        Objects.requireNonNull(value, "text");
        var normalized = Normalizer.normalize(value, Normalizer.Form.NFC);
        var output = new StringBuilder();
        var pendingSpace = false;
        for (var offset = 0; offset < normalized.length();) {
            var codePoint = normalized.codePointAt(offset);
            offset += Character.charCount(codePoint);
            if (Character.isISOControl(codePoint)) {
                throw invalid("R5P2_RECONCILIATION_TEXT_INVALID");
            }
            if (Character.isWhitespace(codePoint) || Character.isSpaceChar(codePoint)) {
                pendingSpace = output.length() > 0;
            } else {
                if (pendingSpace) output.append(' ');
                output.appendCodePoint(codePoint);
                pendingSpace = false;
            }
        }
        if (output.isEmpty() || !output.toString().equals(value)) {
            throw invalid("R5P2_RECONCILIATION_TEXT_INVALID");
        }
        return value;
    }

    private static String localId(String value) {
        if (value == null || !value.matches("[a-z][a-z0-9-]{0,127}")) {
            throw invalid("R5P2_RECONCILIATION_OBSERVATION_ID_INVALID");
        }
        return value;
    }

    private static String artifactId(String value) {
        if (value == null || !value.matches("[0-9a-f]{64}")) {
            throw invalid("R5P2_RECONCILIATION_SOURCE_ID_INVALID");
        }
        return value;
    }

    private static String sha256(byte[] value) {
        try {
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(value));
        } catch (NoSuchAlgorithmException impossible) {
            throw new IllegalStateException("SHA-256 is required", impossible);
        }
    }

    private static IllegalArgumentException invalid(String code) {
        return new IllegalArgumentException(code);
    }

    public record PixelBox(int left, int top, int right, int bottom) {
        public PixelBox {
            if (left < 0 || top < 0 || left >= right || top >= bottom) {
                throw invalid("R5P2_PROJECTION_PIXEL_BOX_INVALID");
            }
        }

        void requireWithin(int width, int height) {
            if (width < 1 || height < 1 || right > width || bottom > height) {
                throw invalid("R5P2_PROJECTION_PIXEL_BOX_INVALID");
            }
        }
    }

    public record SourceBox(int left, int top, int right, int bottom) {
        public SourceBox {
            if (left < 0 || top < 0 || left >= right || top >= bottom
                    || right > 10_000 || bottom > 10_000) {
                throw invalid("R5P2_RECONCILIATION_SOURCE_BOX_INVALID");
            }
        }

        long area() {
            return Math.multiplyExact((long) right - left, bottom - top);
        }

        int height() {
            return bottom - top;
        }
    }

    public record PixelDensity(long numerator, long denominator)
            implements Comparable<PixelDensity> {
        public PixelDensity {
            if (numerator < 1 || denominator < 1) {
                throw invalid("R5P2_RECONCILIATION_DENSITY_INVALID");
            }
        }

        @Override
        public int compareTo(PixelDensity other) {
            Objects.requireNonNull(other, "other");
            return BigInteger.valueOf(numerator).multiply(BigInteger.valueOf(other.denominator))
                    .compareTo(BigInteger.valueOf(other.numerator)
                            .multiply(BigInteger.valueOf(denominator)));
        }
    }

    public record ProjectedLine(
            String observationId,
            String sourceArtifactId,
            SourceBox sourceBox,
            int confidenceBps,
            String text,
            int viewOrdinal,
            int lineOrdinal,
            PixelDensity pixelDensity
    ) {
        public ProjectedLine {
            observationId = localId(observationId);
            sourceArtifactId = artifactId(sourceArtifactId);
            Objects.requireNonNull(sourceBox, "sourceBox");
            if (confidenceBps < 0 || confidenceBps > 10_000
                    || viewOrdinal < 0 || viewOrdinal >= 10
                    || lineOrdinal < 0 || lineOrdinal >= 4_096) {
                throw invalid("R5P2_RECONCILIATION_LINE_BOUNDS_INVALID");
            }
            text = canonicalText(text);
            Objects.requireNonNull(pixelDensity, "pixelDensity");
        }
    }

    public record Outcome(
            List<ProjectedLine> representatives,
            int inputCount,
            int clusterCount,
            String policyIdentity
    ) {
        public Outcome {
            representatives = List.copyOf(Objects.requireNonNull(representatives, "representatives"));
            if (inputCount < 1 || clusterCount != representatives.size()
                    || clusterCount < 1 || !POLICY_IDENTITY.equals(policyIdentity)) {
                throw invalid("R5P2_RECONCILIATION_OUTCOME_INVALID");
            }
        }
    }
}

File: tools/r5p2_source_line_reconciliation.py
#!/usr/bin/env python3
"""Independent Python implementation of the frozen R5P2 reconciliation policy."""

from __future__ import annotations

import hashlib
import re
import unicodedata
from dataclasses import dataclass
from typing import Any, Sequence


VERSION = "FrozenSourceLineReconciliationPolicy/1.0"
PROJECTION_IDENTITY = "renderweave-r5p-source-projection/1.0"
AREA_OVERLAP_BPS = 5_000
VERTICAL_OVERLAP_BPS = 8_000
POLICY_FRAMES = (
    VERSION,
    f"projection={PROJECTION_IDENTITY}",
    "scope=same-source-cross-view-only/1.0",
    "geometry=intersection-over-smaller-area-5000-bps/1.0",
    "vertical=intersection-over-smaller-height-8000-bps/1.0",
    "center=smaller-center-in-larger-closed-open/1.0",
    "cluster=complete-link-source-order/1.0",
    "representative=pixel-density-confidence-smaller-area-view-line/1.0",
    "representative-payload=original-observation-only/1.0",
    "canonical-text-order=unicode-scalar-lexicographic/1.0",
    "sort=top-left-bottom-right-canonical-text-view-line/1.0",
)
POLICY_IDENTITY = f"{VERSION}:" + hashlib.sha256("\n".join(POLICY_FRAMES).encode()).hexdigest()
if POLICY_IDENTITY != (
    "FrozenSourceLineReconciliationPolicy/1.0:"
    "eead9287d942693156500a090daf5da5c2f9dafe4f6564ee642ac406c0f49443"
):
    raise RuntimeError("R5P2_RECONCILIATION_POLICY_IDENTITY_DRIFT")


class ReconciliationError(ValueError):
    pass


def fail(code: str) -> None:
    raise ReconciliationError(code)


@dataclass(frozen=True)
class PixelBox:
    left: int
    top: int
    right: int
    bottom: int

    def __post_init__(self) -> None:
        if (any(type(value) is not int for value in (self.left, self.top, self.right, self.bottom))
                or self.left < 0 or self.top < 0
                or self.left >= self.right or self.top >= self.bottom):
            fail("R5P2_PROJECTION_PIXEL_BOX_INVALID")

    def require_within(self, width: int, height: int) -> None:
        if (type(width) is not int or type(height) is not int or width < 1 or height < 1
                or self.right > width or self.bottom > height):
            fail("R5P2_PROJECTION_PIXEL_BOX_INVALID")


@dataclass(frozen=True)
class SourceBox:
    left: int
    top: int
    right: int
    bottom: int

    def __post_init__(self) -> None:
        if (any(type(value) is not int for value in (self.left, self.top, self.right, self.bottom))
                or self.left < 0 or self.top < 0 or self.left >= self.right
                or self.top >= self.bottom or self.right > 10_000 or self.bottom > 10_000):
            fail("R5P2_RECONCILIATION_SOURCE_BOX_INVALID")

    @property
    def area(self) -> int:
        return (self.right - self.left) * (self.bottom - self.top)

    @property
    def height(self) -> int:
        return self.bottom - self.top


@dataclass(frozen=True)
class PixelDensity:
    numerator: int
    denominator: int

    def __post_init__(self) -> None:
        if (type(self.numerator) is not int or type(self.denominator) is not int
                or self.numerator < 1 or self.denominator < 1):
            fail("R5P2_RECONCILIATION_DENSITY_INVALID")


def _canonical_text(value: Any) -> str:
    if type(value) is not str or not value:
        fail("R5P2_RECONCILIATION_TEXT_INVALID")
    normalized = unicodedata.normalize("NFC", value)
    if any(unicodedata.category(character) == "Cc" for character in normalized):
        fail("R5P2_RECONCILIATION_TEXT_INVALID")
    canonical = " ".join(normalized.split())
    if value != canonical:
        fail("R5P2_RECONCILIATION_TEXT_INVALID")
    return value


@dataclass(frozen=True)
class ProjectedLine:
    observation_id: str
    source_artifact_id: str
    source_box: SourceBox
    confidence_bps: int
    text: str
    view_ordinal: int
    line_ordinal: int
    pixel_density: PixelDensity

    def __post_init__(self) -> None:
        if type(self.observation_id) is not str or not re.fullmatch(r"[a-z][a-z0-9-]{0,127}", self.observation_id):
            fail("R5P2_RECONCILIATION_OBSERVATION_ID_INVALID")
        if type(self.source_artifact_id) is not str or not re.fullmatch(r"[0-9a-f]{64}", self.source_artifact_id):
            fail("R5P2_RECONCILIATION_SOURCE_ID_INVALID")
        if (type(self.source_box) is not SourceBox or type(self.pixel_density) is not PixelDensity
                or type(self.confidence_bps) is not int or not 0 <= self.confidence_bps <= 10_000
                or type(self.view_ordinal) is not int or not 0 <= self.view_ordinal < 10
                or type(self.line_ordinal) is not int or not 0 <= self.line_ordinal < 4_096):
            fail("R5P2_RECONCILIATION_LINE_BOUNDS_INVALID")
        _canonical_text(self.text)


@dataclass(frozen=True)
class Outcome:
    representatives: tuple[ProjectedLine, ...]
    input_count: int
    cluster_count: int
    policy_identity: str = POLICY_IDENTITY

    def __post_init__(self) -> None:
        if (not self.representatives or self.input_count < 1
                or self.cluster_count != len(self.representatives)
                or self.policy_identity != POLICY_IDENTITY):
            fail("R5P2_RECONCILIATION_OUTCOME_INVALID")


def _floor_ratio(value: int, multiplier: int, divisor: int) -> int:
    if divisor < 1:
        fail("R5P2_PROJECTION_DIMENSIONS_INVALID")
    return value * multiplier // divisor


def _ceil_ratio(value: int, multiplier: int, divisor: int) -> int:
    if divisor < 1:
        fail("R5P2_PROJECTION_DIMENSIONS_INVALID")
    return (value * multiplier + divisor - 1) // divisor


def _project_floor(crop_start: int, crop_size: int, coordinate: int, source_size: int) -> int:
    if source_size < 1:
        fail("R5P2_PROJECTION_DIMENSIONS_INVALID")
    return (crop_start * 10_000 + coordinate * crop_size) // source_size


def _project_ceil(crop_start: int, crop_size: int, coordinate: int, source_size: int) -> int:
    if source_size < 1:
        fail("R5P2_PROJECTION_DIMENSIONS_INVALID")
    numerator = crop_start * 10_000 + coordinate * crop_size
    return (numerator + source_size - 1) // source_size


def project(
        observation_id: str, source_artifact_id: str, view_ordinal: int, line_ordinal: int,
        view_width: int, view_height: int, source_width: int, source_height: int,
        source_crop: PixelBox, view_line: PixelBox, confidence_bps: int,
        text: str) -> ProjectedLine:
    source_crop.require_within(source_width, source_height)
    view_line.require_within(view_width, view_height)
    view_canonical = SourceBox(
        _floor_ratio(view_line.left, 10_000, view_width),
        _floor_ratio(view_line.top, 10_000, view_height),
        _ceil_ratio(view_line.right, 10_000, view_width),
        _ceil_ratio(view_line.bottom, 10_000, view_height),
    )
    crop_width = source_crop.right - source_crop.left
    crop_height = source_crop.bottom - source_crop.top
    source_box = SourceBox(
        _project_floor(source_crop.left, crop_width, view_canonical.left, source_width),
        _project_floor(source_crop.top, crop_height, view_canonical.top, source_height),
        _project_ceil(source_crop.left, crop_width, view_canonical.right, source_width),
        _project_ceil(source_crop.top, crop_height, view_canonical.bottom, source_height),
    )
    return ProjectedLine(
        observation_id, source_artifact_id, source_box, confidence_bps, text,
        view_ordinal, line_ordinal,
        PixelDensity(view_width * view_height, crop_width * crop_height),
    )


def _at_least_bps(numerator: int, denominator: int, threshold: int) -> bool:
    return denominator > 0 and numerator * 10_000 >= denominator * threshold


def area_threshold_allows(intersection_area: int, smaller_area: int) -> bool:
    return _at_least_bps(intersection_area, smaller_area, AREA_OVERLAP_BPS)


def vertical_threshold_allows(intersection_height: int, smaller_height: int) -> bool:
    return _at_least_bps(intersection_height, smaller_height, VERTICAL_OVERLAP_BPS)


def _center_inside(smaller: SourceBox, larger: SourceBox) -> bool:
    center_x2 = smaller.left + smaller.right
    center_y2 = smaller.top + smaller.bottom
    return (larger.left * 2 <= center_x2 < larger.right * 2
            and larger.top * 2 <= center_y2 < larger.bottom * 2)


def same_source_line_candidate(left: ProjectedLine, right: ProjectedLine) -> bool:
    if left.source_artifact_id != right.source_artifact_id or left.view_ordinal == right.view_ordinal:
        return False
    intersection_width = max(
        0, min(left.source_box.right, right.source_box.right)
        - max(left.source_box.left, right.source_box.left),
    )
    intersection_height = max(
        0, min(left.source_box.bottom, right.source_box.bottom)
        - max(left.source_box.top, right.source_box.top),
    )
    intersection = intersection_width * intersection_height
    if (intersection == 0
            or not area_threshold_allows(
                intersection, min(left.source_box.area, right.source_box.area))
            or not vertical_threshold_allows(
                intersection_height, min(left.source_box.height, right.source_box.height))):
        return False
    if left.source_box.area < right.source_box.area:
        return _center_inside(left.source_box, right.source_box)
    if right.source_box.area < left.source_box.area:
        return _center_inside(right.source_box, left.source_box)
    return (_center_inside(left.source_box, right.source_box)
            and _center_inside(right.source_box, left.source_box))


def _source_order(value: ProjectedLine) -> tuple[Any, ...]:
    box = value.source_box
    return (box.top, box.left, box.bottom, box.right, value.text, value.view_ordinal,
            value.line_ordinal)


def prefers_representative(candidate: ProjectedLine, existing: ProjectedLine) -> bool:
    left = candidate.pixel_density.numerator * existing.pixel_density.denominator
    right = existing.pixel_density.numerator * candidate.pixel_density.denominator
    if left != right:
        return left > right
    if candidate.confidence_bps != existing.confidence_bps:
        return candidate.confidence_bps > existing.confidence_bps
    if candidate.source_box.area != existing.source_box.area:
        return candidate.source_box.area < existing.source_box.area
    if candidate.view_ordinal != existing.view_ordinal:
        return candidate.view_ordinal < existing.view_ordinal
    return candidate.line_ordinal < existing.line_ordinal


def reconcile(lines: Sequence[ProjectedLine]) -> Outcome:
    lines = tuple(lines)
    if not 1 <= len(lines) <= 4_096 or any(type(line) is not ProjectedLine for line in lines):
        fail("R5P2_RECONCILIATION_INPUT_INVALID")
    if len({line.observation_id for line in lines}) != len(lines):
        fail("R5P2_RECONCILIATION_OBSERVATION_DUPLICATED")
    clusters: list[list[ProjectedLine]] = []
    for candidate in sorted(lines, key=_source_order):
        for cluster in clusters:
            if all(same_source_line_candidate(candidate, existing) for existing in cluster):
                cluster.append(candidate)
                break
        else:
            clusters.append([candidate])
    representatives: list[ProjectedLine] = []
    for cluster in clusters:
        representative = cluster[0]
        for candidate in cluster:
            if prefers_representative(candidate, representative):
                representative = candidate
        representatives.append(representative)
    representatives.sort(key=_source_order)
    return Outcome(tuple(representatives), len(lines), len(clusters))


def projected_line(value: dict[str, Any]) -> ProjectedLine:
    expected = {
        "observationId", "sourceArtifactId", "box", "confidenceBps", "text",
        "viewOrdinal", "lineOrdinal", "density",
    }
    if type(value) is not dict or set(value) != expected:
        fail("R5P2_RECONCILIATION_GOLDEN_INVALID")
    box = value["box"]
    density = value["density"]
    if (type(box) is not list or len(box) != 4
            or type(density) is not list or len(density) != 2):
        fail("R5P2_RECONCILIATION_GOLDEN_INVALID")
    return ProjectedLine(
        value["observationId"], value["sourceArtifactId"], SourceBox(*box),
        value["confidenceBps"], value["text"], value["viewOrdinal"],
        value["lineOrdinal"], PixelDensity(*density),
    )
